`Range [0, -1) out of bounds for length 29` in `TranspileMojo`

```
Caused by: java.lang.StringIndexOutOfBoundsException: Range [0, -1) out of bounds for length 29
    at jdk.internal.util.Preconditions$1.apply (Preconditions.java:55)
    at jdk.internal.util.Preconditions$1.apply (Preconditions.java:52)
    at jdk.internal.util.Preconditions$4.apply (Preconditions.java:213)
    at jdk.internal.util.Preconditions$4.apply (Preconditions.java:210)
    at jdk.internal.util.Preconditions.outOfBounds (Preconditions.java:98)
    at jdk.internal.util.Preconditions.outOfBoundsCheckFromToIndex (Preconditions.java:112)
    at jdk.internal.util.Preconditions.checkFromToIndex (Preconditions.java:349)
    at java.lang.String.checkBoundsBeginEnd (String.java:4963)
    at java.lang.String.substring (String.java:2925)
    at org.eolang.maven.TranspileMojo.pinfo (TranspileMojo.java:338)
```

--- a/eo-maven-plugin/src/main/java/org/eolang/maven/TranspileMojo.java
+++ b/eo-maven-plugin/src/main/java/org/eolang/maven/TranspileMojo.java
@@ -38,6 +38,7 @@ import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
+import java.util.Arrays;
 import java.util.Collection;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
@@ -253,7 +254,7 @@ public final class TranspileMojo extends SafeMojo {
             final Path tgt = new Place(jname).make(
                 this.generatedDir.toPath(), TranspileMojo.JAVA
             );
-            this.pinfo(tgt, jname, java.xpath("@package").get(0));
+            this.pinfo(tgt, jname, java.xpath("@package").stream().findFirst().orElse(""));
             final Supplier<Path> che = new CachePath(
                 this.cache.toPath().resolve(TranspileMojo.CACHE),
                 this.plugin.getVersion(),
@@ -335,7 +336,13 @@ public final class TranspileMojo extends SafeMojo {
                     " * don't modify it, all changes will be lost anyway.",
                     " */",
                     String.format("// @org.eolang.XmirPackage(\"%s\")", pname),
-                    String.format("package %s;", oname.substring(0, oname.lastIndexOf('.')))
+                    String.format(
+                        "package %s;",
+                        Arrays.stream(oname.split("\\."))
+                            .reduce((f, s) -> s)
+                            .stream()
+                            .findFirst()
+                    )
                 ).getBytes(StandardCharsets.UTF_8)
             );
             Logger.debug(this, "Saved %[file]s (%[size]s)", pinfo, pinfo.toFile().length());

--- a/eo-maven-plugin/src/main/java/org/eolang/maven/TranspileMojo.java
+++ b/eo-maven-plugin/src/main/java/org/eolang/maven/TranspileMojo.java
@@ -38,7 +38,6 @@ import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
-import java.util.Arrays;
 import java.util.Collection;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicInteger;
@@ -323,9 +322,13 @@ public final class TranspileMojo extends SafeMojo {
     private void pinfo(final Path java, final String oname, final String pname)
         throws IOException {
         final Path pinfo = java.getParent().resolve("package-info.java");
-        if (!pinfo.toFile().exists()) {
+        if (!pinfo.toFile().exists() && !pname.isEmpty()) {
             if (pinfo.getParent().toFile().mkdirs()) {
                 Logger.debug(this, "Directory created for %[file]s", pinfo);
+            }
+            String pkg = oname;
+            if (oname.contains(".")) {
+                pkg = pkg.substring(0, pkg.lastIndexOf('.'));
             }
             Files.write(
                 pinfo,
@@ -336,13 +339,7 @@ public final class TranspileMojo extends SafeMojo {
                     " * don't modify it, all changes will be lost anyway.",
                     " */",
                     String.format("// @org.eolang.XmirPackage(\"%s\")", pname),
-                    String.format(
-                        "package %s;",
-                        Arrays.stream(oname.split("\\."))
-                            .reduce((f, s) -> s)
-                            .stream()
-                            .findFirst()
-                    )
+                    String.format("package %s;", pkg)
                 ).getBytes(StandardCharsets.UTF_8)
             );
             Logger.debug(this, "Saved %[file]s (%[size]s)", pinfo, pinfo.toFile().length());

--- a/eo-maven-plugin/src/test/java/org/eolang/maven/TranspileMojoTest.java
+++ b/eo-maven-plugin/src/test/java/org/eolang/maven/TranspileMojoTest.java
@@ -127,12 +127,12 @@ final class TranspileMojoTest {
     }
 
     @Test
-    void transpilesSimpleApp(@Mktmp final Path temp, @RandomProgram final String program)
+    void transpilesSimpleApp(@Mktmp final Path temp, @RandomProgram final String prog)
         throws Exception {
         new Farea(temp).together(
             f -> {
                 f.clean();
-                f.files().file("src/main/eo/foo.eo").write(program.getBytes());
+                f.files().file("src/main/eo/foo.eo").write(prog.getBytes());
                 f.build()
                     .plugins()
                     .appendItself()

--- a/eo-maven-plugin/src/test/java/org/eolang/maven/TranspileMojoTest.java
+++ b/eo-maven-plugin/src/test/java/org/eolang/maven/TranspileMojoTest.java
@@ -26,7 +26,6 @@ package org.eolang.maven;
 import com.yegor256.Mktmp;
 import com.yegor256.MktmpResolver;
 import com.yegor256.farea.Farea;
-import com.yegor256.farea.RequisiteMatcher;
 import java.io.IOException;
 import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
@@ -108,15 +107,8 @@ final class TranspileMojoTest {
                     .plugins()
                     .appendItself()
                     .execution()
-                    .goals("register", "parse", "optimize", "shake", "transpile")
-                    .configuration()
-                    .set("failOnWarning", "false");
+                    .goals("register", "parse", "optimize", "shake", "transpile");
                 f.exec("process-sources");
-                MatcherAssert.assertThat(
-                    "the build was clean",
-                    f.log(),
-                    RequisiteMatcher.SUCCESS
-                );
             }
         );
         MatcherAssert.assertThat(
@@ -145,15 +137,8 @@ final class TranspileMojoTest {
                     .plugins()
                     .appendItself()
                     .execution()
-                    .goals("register", "parse", "optimize", "shake", "transpile")
-                    .configuration()
-                    .set("failOnWarning", "false");
+                    .goals("register", "parse", "optimize", "shake", "transpile");
                 f.exec("process-sources");
-                MatcherAssert.assertThat(
-                    "the build was clean",
-                    f.log(),
-                    RequisiteMatcher.SUCCESS
-                );
             }
         );
         MatcherAssert.assertThat(

--- a/eo-maven-plugin/src/test/java/org/eolang/maven/TranspileMojoTest.java
+++ b/eo-maven-plugin/src/test/java/org/eolang/maven/TranspileMojoTest.java
@@ -25,7 +25,10 @@ package org.eolang.maven;
 
 import com.yegor256.Mktmp;
 import com.yegor256.MktmpResolver;
+import com.yegor256.farea.Farea;
+import com.yegor256.farea.RequisiteMatcher;
 import java.io.IOException;
+import java.nio.charset.StandardCharsets;
 import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
@@ -54,6 +57,7 @@ import org.junit.jupiter.params.ParameterizedTest;
  */
 @SuppressWarnings({"PMD.AvoidDuplicateLiterals", "PMD.TooManyMethods"})
 @ExtendWith(MktmpResolver.class)
+@ExtendWith(RandomProgramResolver.class)
 final class TranspileMojoTest {
 
     /**
@@ -79,6 +83,88 @@ final class TranspileMojoTest {
             "passes with no exceptions",
             new XtSticky(new XtYaml(yaml)),
             new XtoryMatcher()
+        );
+    }
+
+    @Test
+    void transpilesWithPackage(@Mktmp final Path temp)
+        throws Exception {
+        new Farea(temp).together(
+            f -> {
+                f.clean();
+                f.files().file("src/main/eo/one/foo.eo").write(
+                    String.join(
+                        "\n",
+                        "+package one",
+                        "",
+                        "# no comments.",
+                        "[] > foo",
+                        "  QQ.io.stdout > @",
+                        "    \"Hello, world!\\n\"",
+                        ""
+                    ).getBytes(StandardCharsets.UTF_8)
+                );
+                f.build()
+                    .plugins()
+                    .appendItself()
+                    .execution()
+                    .goals("register", "parse", "optimize", "shake", "transpile")
+                    .configuration()
+                    .set("failOnWarning", "false");
+                f.exec("process-sources");
+                MatcherAssert.assertThat(
+                    "the build was clean",
+                    f.log(),
+                    RequisiteMatcher.SUCCESS
+                );
+            }
+        );
+        MatcherAssert.assertThat(
+            "the .java file is generated",
+            temp.resolve("target/generated-sources/EOone/EOfoo.java").toFile().exists(),
+            Matchers.is(true)
+        );
+        MatcherAssert.assertThat(
+            "the package-info.java file contains the right package name",
+            Files.readString(
+                temp.resolve("target/generated-sources/EOone/package-info.java"),
+                StandardCharsets.UTF_8
+            ),
+            Matchers.containsString("package EOone;")
+        );
+    }
+
+    @Test
+    void transpilesSimpleApp(@Mktmp final Path temp, @RandomProgram final String program)
+        throws Exception {
+        new Farea(temp).together(
+            f -> {
+                f.clean();
+                f.files().file("src/main/eo/foo.eo").write(program.getBytes());
+                f.build()
+                    .plugins()
+                    .appendItself()
+                    .execution()
+                    .goals("register", "parse", "optimize", "shake", "transpile")
+                    .configuration()
+                    .set("failOnWarning", "false");
+                f.exec("process-sources");
+                MatcherAssert.assertThat(
+                    "the build was clean",
+                    f.log(),
+                    RequisiteMatcher.SUCCESS
+                );
+            }
+        );
+        MatcherAssert.assertThat(
+            "the .java file is re-generated",
+            temp.resolve("target/generated-sources/EOfoo.java").toFile().exists(),
+            Matchers.is(true)
+        );
+        MatcherAssert.assertThat(
+            "the package-info.java file contains the right package name",
+            temp.resolve("target/generated-sources/package-info.java").toFile().exists(),
+            Matchers.is(false)
         );
     }
 

